Description: Segfault if a .paths directory is encountered in $PATH
  ksh crashes if it encounters a .paths directory in any of the
  directories in $PATH.
Forwarded: yes
Author: Anuradha Weeraman <anuradha@debian.org>
Reviewed-By: Martijn Dekker <martijn@inlv.org>
Bug-Ubuntu: https://bugs.launchpad.net/ubuntu/+source/ksh/+bug/1534855
Origin: https://github.com/ksh93/ksh/commit/035a4cb3f453271b7ae63bcb53a7963b8dbe4c41
Last-Update: 2020-07-02

diff --git a/src/cmd/ksh93/sh/path.c b/src/cmd/ksh93/sh/path.c
index 92cc1ea8..356d5954 100644
--- a/src/cmd/ksh93/sh/path.c
+++ b/src/cmd/ksh93/sh/path.c
@@ -1510,6 +1510,10 @@ static int path_chkpaths(Shell_t *shp,Pathcomp_t *first, Pathcomp_t* old,Pathcom
 	if((fd=open(stakptr(offset),O_RDONLY))>=0)
 	{
 		fstat(fd,&statb);
+		if (!S_ISREG(statb.st_mode)) {
+			close(fd);
+			return 0;
+		}
 		n = statb.st_size;
 		stakseek(offset+pp->len+n+2);
 		sp = stakptr(offset+pp->len);
diff --git a/src/cmd/ksh93/tests/path.sh b/src/cmd/ksh93/tests/path.sh
index 491de73d..0b4b8d72 100755
--- a/src/cmd/ksh93/tests/path.sh
+++ b/src/cmd/ksh93/tests/path.sh
@@ -391,5 +391,14 @@ ${.sh.version}
 END
 ) || err_exit '${.sh.xxx} variables causes cat not be found'
 
+# ksh segfaults if $PATH contains a .paths directory
+mkdir -p $tmp/paths-dir-crash/
+cat > $tmp/paths-dir-crash/run.sh <<- EOF
+mkdir -p $tmp/paths-dir-crash/.paths
+export PATH=$tmp/paths-dir-crash:$PATH
+print ok
+EOF
+[[ $($SHELL $tmp/paths-dir-crash/run.sh 2>/dev/null) == ok ]] || err_exit "ksh crashes if PATH contains a .paths directory"
+
 exit $((Errors<125?Errors:125))
 
